cpp/ql/test/library-tests/dataflow/dataflow-tests/has-parameter-flow-out.ql

```diff
@@ -28,6 +28,11 @@ module IRTest {
   private import semmle.code.cpp.ir.dataflow.DataFlow
   private import semmle.code.cpp.ir.dataflow.internal.DataFlowUtil
 
+  private string stars(int k) {
+    k = [0 .. max(FinalParameterNode n | | n.getIndirectionIndex())] and
```

Comment:
Just for my understanding: this is needed because `k` needs to be bound somehow?

Reply:
Yeah, exactly. Remember that `if k = 0 then result = "" else result = "*" + stars(k - 1)` desugars to something like:
```ql
k = 0 and result = ""
or
not k = 0 and result = "*" + stars(k - 1)
```
and, while the first disjunct binds `k`, the second discount doesn't.

Normally, we could just add a `bindingset[k]` annotation on the predicate, but we can't do that for recursive predicates (since `bindingset` effectively means `inline`). So we have to bind `k` inside the predicate.

Reply:
Thanks!